pkg/compute/executor.go

```diff
@@ -86,141 +121,220 @@ func PrepareRunArguments(
 		(@wdbaruni's comment: https://github.com/bacalhau-project/bacalhau/pull/2637#issuecomment-1625739030
 		provides more context on the need for the change).
 	*/
+	var engineArgs *models.SpecConfig
 	if execution.Job.Task().Engine.IsType(models.EngineWasm) {
 		wasmEngine, err := wasmmodels.DecodeSpec(execution.Job.Task().Engine)
 		if err != nil {
-			return nil, err
-		}
-		importModuleVolumes, err := storage.ParallelPrepareStorage(ctx, strgprovider, wasmEngine.ImportModules...)
-		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-		cleanup.RegisterCallbackWithContext(func(ctx context.Context) error {
-			if err := storage.ParallelCleanStorage(ctx, strgprovider, importModuleVolumes); err != nil {
-				return fmt.Errorf("cleaning up wasm import modules: %w", err)
-			}
-			return nil
-		})
 
-		entryModuleVolumes, err := storage.ParallelPrepareStorage(ctx, strgprovider, wasmEngine.EntryModule)
+		volumes, wasmCleanup, err := prepareWasmVolumes(ctx, strgprovider, wasmEngine)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-		cleanup.RegisterCallbackWithContext(func(ctx context.Context) error {
-			if err := storage.ParallelCleanStorage(ctx, strgprovider, entryModuleVolumes); err != nil {
-				return fmt.Errorf("cleaning up wasm entry modules: %w", err)
-			}
-			return nil
-		})
+
+		cleanupFuncs = append(cleanupFuncs, wasmCleanup)
+
 		engineArgs = &models.SpecConfig{
 			Type:   models.EngineWasm,
-			Params: wasmEngine.ToArguments(entryModuleVolumes[0], importModuleVolumes...).ToMap(),
+			Params: wasmEngine.ToArguments(volumes["entryModules"][0], volumes["importModules"]...).ToMap(),
 		}
 	} else {
 		engineArgs = execution.Job.Task().Engine
 	}
+
 	return &executor.RunCommandRequest{
-		JobID:        execution.Job.ID,
-		ExecutionID:  execution.ID,
-		Resources:    execution.TotalAllocatedResources(),
-		Network:      execution.Job.Task().Network,
-		Outputs:      execution.Job.Task().ResultPaths,
-		Inputs:       inputVolumes,
-		ResultsDir:   resultsDir,
-		EngineParams: engineArgs,
-		OutputLimits: executor.OutputLimits{
-			MaxStdoutFileLength:   system.MaxStdoutFileLength,
-			MaxStdoutReturnLength: system.MaxStdoutReturnLength,
-			MaxStderrFileLength:   system.MaxStderrFileLength,
-			MaxStderrReturnLength: system.MaxStderrReturnLength,
-		},
-	}, nil
+			JobID:        execution.Job.ID,
+			ExecutionID:  execution.ID,
+			Resources:    execution.TotalAllocatedResources(),
+			Network:      execution.Job.Task().Network,
+			Outputs:      execution.Job.Task().ResultPaths,
+			Inputs:       inputVolumes,
+			ResultsDir:   resultsDir,
+			EngineParams: engineArgs,
+			OutputLimits: executor.OutputLimits{
+				MaxStdoutFileLength:   system.MaxStdoutFileLength,
+				MaxStdoutReturnLength: system.MaxStdoutReturnLength,
+				MaxStderrFileLength:   system.MaxStderrFileLength,
+				MaxStderrReturnLength: system.MaxStderrReturnLength,
+			},
+		}, func(ctx context.Context) error {
```

Comment:
Might it be worth moving this cleanup function into the `RunCommandRequest`?  At the moment it's an anonymous looking func and it isn't clear in the signature what it is/what it's for.  By adding it to the request we can turn it into a post-execute/cleanup step, it's clearer what it is.  Understand if you'd like to keep it outside the structure, but in that case it would be good to explain what it is in the func doc/comment.

Reply:
Agreed, the current code structure is lacking comments describing it's usage, I'll add them before taking this out of draft.

> Might it be worth moving this cleanup function into the `RunCommandRequest`?

I think adding the cleanup to a structure related to the bits being cleaned up would be cleaner, so generally in favor of this.  However, I think it might also fit on the `models.Execution`, which would allow us to perform the cleanup when the execution reaches a terminal state. Since the state we are cleaning here is related to inputs we could probably (safely?) keep it around in the event that an execution is rescheduled or restarted. What do you think?

For now I will keep the cleanup logic as a clearly-commented anonymous function.

Reply:
I think I'd rather leave it where it is than add it to the Execution as it's a pretty inactive model at present and intended to be immutable. I don't think it's definitively wrong, it just feels weird returning a 3-tuple from a method instead of the usual 2-tuple.